### PR TITLE
PADV-1190: Add Deep Linking views.

### PR DESCRIPTION
**Pull request by kuipumu**
## Ticket

https://agile-jira.pearson.com/browse/PADV-1190

## Description

This PR adds the views for the Deep Linking feature, currently the view can be interacted with a Deep Linking request but there is still some development pending to be done on these views that will be addressed on PADV-1191 that will allow to use the views on a complete Deep Linking UI workflow. 

## Type of Change

- [x] Add `deep_linking` module.
- [x] Add `DeepLinkingView` class.
- [x] Add `DeepLinkingFormView` class.
- [x] Add `http_response_error` method to `LtiToolBaseView` view.
- [x] Add or modify all tests related to this code.

## Testing:

- Run the LMS: `make dev.up.lms`.
- Install `openedx_lti_tool_plugin` on the LMS.
- Run ngrok or any other similar service on LMS: `ngrok http 18000`
- Add required settings to LMS.
	
```
# Enable LTI Tool Provider Plugin.
OLTITP_ENABLE_LTI_TOOL = True
```

- Create a new LTI 1.3 tool key config: http://localhost:18000/admin/lti1p3_tool_config/ltitoolkey/add/ (You can use IMS RI to create a key pair: https://lti-ri.imsglobal.org/keygen/index)
- Create a new LTI 1.3 tool config: http://localhost:18000/admin/lti1p3_tool_config/ltitool/add/

```
Issuer: https://saltire.lti.app/platform
Client id: saltire.lti.app
Auth login: url: https://saltire.lti.app/platform/auth
Auth token url: https://saltire.lti.app/platform/token/sda42bb0cf2352259a367a404d48d54e8
Key set url: https://saltire.lti.app/platform/jwks/sda42bb0cf2352259a367a404d48d54e8
Deployment ids: ["cLWwj9cbmkSrCNsckEFBmA"]
```

- Go to the saLTIre platform https://saltire.lti.app/platform
- Go to "Security Model" on the left sidebar and set these settings:

```
Message URL: https://lms.ngrok.io/openedx_lti_tool_plugin/1.3/launch/{course_id}
Initiate login URL: https://lms.ngrok.io/openedx_lti_tool_plugin/1.3/login
Redirection URI(s): https://lms.ngrok.io/openedx_lti_tool_plugin/1.3/launch/{course_id}
Public keyset URL: https://lms.ngrok.io/openedx_lti_tool_plugin/1.3/pub/jwks
```

- Go to "Message" on the left sidebar and set these settings:

```
Message type: LtiDeepLinkingRequest
```

- Click on the "Save" button on the top navbar.
- Click on the dropdown next to the "Connect" button on the top navbar.
- Click on "Open in iframe" or "Open in new window".
- The request should take you to the DeepLinkingFormView GET response after being redirected from the DeepLinkingView POST action.

## Reviewers

- [x] @alexjmpb  
- [x] @anfbermudezme

### CODE REVIEW

File: openedx_lti_tool_plugin/deep_linking/views.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        Returns:
+            HttpResponse or LoggedHttpResponseBadRequest.
+
+        """
+        try:
+            message = DjangoMessageLaunch(
+                request,
+                self.tool_config,
+                launch_data_storage=self.tool_storage,
+            )
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/deep_linking/views.py:55]: @kuipumu Add blank line.

File: openedx_lti_tool_plugin/deep_linking/views.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+        Returns:
+            HttpResponse or LoggedHttpResponseBadRequest.
+
+        """
+        try:
+            message = self.get_message_from_cache(request, launch_id)
+            validate_message(message)
+
+            # TODO: Add code to obtain form and validate form with resource link data.
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/deep_linking/views.py:130]: @kuipumu Is this reported in a ticket?

File: openedx_lti_tool_plugin/deep_linking/views.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+            # TODO: Add code to process form cleaned data and generate the Deep Linking response:
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/deep_linking/views.py:131]: @kuipumu I think you can modify this in one TODO.

File: openedx_lti_tool_plugin/deep_linking/views.py [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+
+        Returns:
+            HttpResponse or LoggedHttpResponseBadRequest.
+
+        """
+        try:
+            message = self.get_message_from_cache(request, launch_id)
+            validate_message(message)
+
+            # TODO: Add template response with form to allow user to select
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/deep_linking/views.py:102]: @kuipumu Is this reported in a ticket?

File: openedx_lti_tool_plugin/deep_linking/views.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+        """
+        try:
+            message = self.get_message_from_cache(request, launch_id)
+            validate_message(message)
+
+            # TODO: Add code to obtain form and validate form with resource link data.
+            # TODO: Add code to process form cleaned data and generate the Deep Linking response:
+            # https://github.com/dmitry-viskov/pylti1.3?tab=readme-ov-file#deep-linking-responses
+            return HttpResponse()
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/deep_linking/views.py:133]: @kuipumu What do you think about add LOG 's to identify if the methods are working?

**Review by anfbermudezme — CHANGES_REQUESTED** [review]

File: openedx_lti_tool_plugin/deep_linking/views.py (hunk 1 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/deep_linking/views.py:55]: @anfbermudezme Why do you think a blank line should be added?, should I also add it to line 99 and 127?

File: openedx_lti_tool_plugin/deep_linking/views.py (hunk 2 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/deep_linking/views.py:130]: @anfbermudezme I just updated the both PADV-1190 and PADV-1191 so this dependency between these two tickets is clearer

File: openedx_lti_tool_plugin/deep_linking/views.py (hunk 4 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/deep_linking/views.py:102]: @anfbermudezme I included in the ticket a description that this form will be implemented in PADV-1191

File: openedx_lti_tool_plugin/deep_linking/views.py (hunk 5 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/deep_linking/views.py:133]: @anfbermudezme I don't think is necessary since there could be to many logs for each deep linking request and I don't find much use on such data right now, I will however log each error once the form logic is added and all errors are currently logged with the http_response_error method.

File: openedx_lti_tool_plugin/deep_linking/views.py (hunk 1 above)
Comment (anfbermudezme) [reply openedx_lti_tool_plugin/deep_linking/views.py:55]: @kuipumu To separate variables and methods. It's just a suggestion, I will be agree with your choice.

File: openedx_lti_tool_plugin/tests/test_views.py [hunk 6]
```diff
@@ -22,6 +23,48 @@ def setUp(self):
...
+        self.assertEqual(instance.tool_storage, tool_storage_mock())
+
+    @patch(f'{MODULE_PATH}.LoggedHttpResponseBadRequest')
+    def test_http_response_error(self, http_response_error_mock: MagicMock):
+        """Test `http_response_error` method with error message."""
+        message = 'Error message'
+        instance = self.view_class()
+
+        result = instance.http_response_error(message)
+
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/tests/test_views.py:63]: @kuipumu
```suggestion
self.assertEqual(instance.http_response_error(message), http_response_error_mock())
```

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+        message_launch_mock: MagicMock,
+        tool_storage_mock: MagicMock,
+        tool_conf_mock: MagicMock,
+    ):
+        """Test with the launch_id of a cached DjangoMessageLaunch object."""
+        request = MagicMock()
+        launch_id = uuid4()
+
+        result = DeepLinkingFormView().get_message_from_cache(request, launch_id)
+
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/deep_linking/tests/test_views.py:261]: @kuipumu I think you can call the method inside the assertEqual and even validate the calls.
```suggestion
self.assertEqual(
    DeepLinkingFormView().get_message_from_cache(request, launch_id), 
    message_launch_mock.from_cache(),
)
```

**Review by anfbermudezme — CHANGES_REQUESTED** [review]

**Review by anfbermudezme — APPROVED** [review]

File: openedx_lti_tool_plugin/views.py [hunk 8]
```diff
@@ -55,11 +64,28 @@ def setup(self, request: HttpRequest, *args: tuple, **kwargs: dict):
...
+        Args:
+            message: Error message string or Exception object.
+
+        Returns:
+            LoggedHttpResponseBadRequest object with error message
+                prefixed with LTI version and view name.
+
+        """
+        return LoggedHttpResponseBadRequest(f'LTI 1.3 {self.__class__.__name__}: {message}')
+
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/views.py:88]: What do you think of moving this to it's own commit, separate from deep linking implementation?

File: openedx_lti_tool_plugin/deep_linking/validators.py [hunk 9]
```diff
@@ -0,0 +1,25 @@
+"""Validators."""
+from django.utils.translation import gettext as _
+from pylti1p3.contrib.django import DjangoMessageLaunch
+
+from openedx_lti_tool_plugin.deep_linking.exceptions import DeepLinkingException
+
+
+def validate_message(message: DjangoMessageLaunch):
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/deep_linking/validators.py:8]: What do you think of being more explicit with the name for the function?

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+    def setUp(self):
+        """Set up test fixtures."""
+        super().setUp()
+        self.view_class = DeepLinkingView
+        self.factory = RequestFactory()
+        self.url = reverse('1.3:deep-linking:root')
+        self.request = self.factory.post(self.url)
+
+    def test_with_deep_linking_request(
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/deep_linking/tests/test_views.py:33]: Could you please rename this case so it is clear that it's the happy path? Or specify it in the docstring.

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py [hunk 11]
```diff
@@ -0,0 +1,277 @@
...
+        validate_message_mock.assert_called_once_with(message_launch_mock())
+        message_launch_mock().get_launch_id.assert_called_once_with()
+        message_launch_mock().get_launch_id().replace.assert_called_once_with('lti1p3-launch-', '')
+        redirect_mock.assert_called_once_with(
+            f'{app_config.name}:1.3:deep-linking:form',
+            launch_id=message_launch_mock().get_launch_id().replace(),
+        )
+
+    @patch.object(DeepLinkingView, 'http_response_error')
+    def test_with_lti_exception(
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/deep_linking/tests/test_views.py:60]:
```suggestion
    def test_raises_lti_exception(
```
What do you think?

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+            self.request,
+            tool_conf_mock(),
+            launch_data_storage=tool_storage_mock(),
+        )
+        validate_message_mock.assert_not_called()
+        redirect_mock.assert_not_called()
+        http_response_error_mock.assert_called_once_with(exception)
+
+    @patch.object(DeepLinkingView, 'http_response_error')
+    def test_with_deep_linking_exception(
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/deep_linking/tests/test_views.py:87]: Same as last comment about name change.
```suggestion
    def test_with_deep_linking_exception(
```

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+        self.assertEqual(
+            self.view_class.as_view()(self.request, self.launch_id),
+            http_response_mock.return_value,
+        )
+        get_message_from_cache_mock.assert_called_once_with(self.request, self.launch_id)
+        validate_message_mock.assert_called_once_with(get_message_from_cache_mock())
+        http_response_mock.assert_called_once_with()
+
+    @patch.object(DeepLinkingFormView, 'http_response_error')
+    def test_with_lti_exception(
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/deep_linking/tests/test_views.py:145]: Same comment about name change.

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+        self.assertEqual(
+            self.view_class.as_view()(self.request, self.launch_id),
+            http_response_error_mock.return_value,
+        )
+        get_message_from_cache_mock.assert_called_once_with(self.request, self.launch_id)
+        validate_message_mock.assert_not_called()
+        http_response_error_mock.assert_called_once_with(exception)
+
+    @patch.object(DeepLinkingFormView, 'http_response_error')
+    def test_with_deep_linking_exception(
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/deep_linking/tests/test_views.py:164]: Same comment about name change.

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/deep_linking/tests/test_views.py:233]: Same comment about name change.

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+        self.assertEqual(
+            self.view_class.as_view()(self.request, self.launch_id),
+            http_response_mock.return_value,
+        )
+        get_message_from_cache_mock.assert_called_once_with(self.request, self.launch_id)
+        validate_message_mock.assert_called_once_with(get_message_from_cache_mock())
+        http_response_mock.assert_called_once_with()
+
+    @patch.object(DeepLinkingFormView, 'http_response_error')
+    def test_with_lti_exception(
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/deep_linking/tests/test_views.py:214]: Same comment about name change.

File: openedx_lti_tool_plugin/deep_linking/tests/test_urls.py [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+        """Test view URL."""
+        self.assertEqual(
+            resolve(
+                reverse(
+                    '1.3:deep-linking:form',
+                    kwargs={'launch_id': uuid4()},
+                ),
+            ).func.view_class,
+            DeepLinkingFormView,
+        )
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/deep_linking/tests/test_urls.py:36]: @alexjmpb Because we can make sure the pattern can be resolved properly and that the kwargs that will be sent to the view are correctly named on the urlpattern

File: openedx_lti_tool_plugin/views.py (hunk 8 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/views.py:88]: @alexjmpb I think it's not necessary to separate this on a commit in this PR, and I think opening a PR for this change will be a bigger effort than just including this change on this PR

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py (hunk 10 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/deep_linking/tests/test_views.py:33]: @alexjmpb How do you suggest it should be renamed or how could be specified on the docstring?

File: openedx_lti_tool_plugin/deep_linking/tests/test_urls.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (alexjmpb) [reply openedx_lti_tool_plugin/deep_linking/tests/test_urls.py:36]: But isn't resolving the urls already being tested by Django? Since it's provided by Django I think we can omit the testing.

File: openedx_lti_tool_plugin/deep_linking/tests/test_views.py (hunk 10 above)
Comment (alexjmpb) [reply openedx_lti_tool_plugin/deep_linking/tests/test_views.py:66]: I would say it can be renamed to test_post or putting in the docstring a description that clarifies that this case tests that the post method is being performed correctly.

File: openedx_lti_tool_plugin/deep_linking/tests/test_urls.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/deep_linking/tests/test_urls.py:36]: @alexjmpb This is not testing that the URL resolving works but rather that the declaration of the URL pattern is correct, for example I could unit test a view that calls reverse with X pattern and it will always work since it's a mock that's just testing the call, with this we make sure that such pattern is properly declared on urls.py

**Review by alexjmpb — APPROVED** [review]
